### ISSUE TITLE
How to create a search for table? 

### ISSUE THREAD

**Issue by lavezzi1** [gurghet/vue-smart-table#2]
Hello, @gurghet thanks for the plugin.

I just want to know how to create a search box for table, something like this http://vuejs.org/examples/grid-component.html

Thanks!

**Comment by gurghet** [comment]
As you can see from the roadmap, filtering is to be implemented in the next days.
If you can't wait you can download the [/dev branch](https://github.com/gurghet/vue-smart-table/tree/dev). There is a filter like the one you requested already implemented. Just pay attention to the lines in the index.html
`15: <input type="text" v-model="search"/>`
`32   :can-filter-by="['name.first', 'name.last']"`
and in main.js
`data: {
    search: ''
  },
  watch: {
    'search' (val) {
      this.$broadcast('filterAll', val)
    }
  },`
So basically you send a `filterAll` event with the text you want to filter. Just note that you have to specify the searchable columns with `:can-filter-by="[<array of cols>]"`
You can't filter by columns that are not simple strings for now.
I hope I answered your question.

**Comment by lavezzi1** [comment]
@gurghet 
Thanks. It works but not for all cols. I get this error:
`[Vue warn]: Error when evaluating expression "function processedSmartBody() {
      var _this6 = this;

      function getDataFromDotNotation(d, row) {
        return d.split('.').reduce(function (o, i) {
          return o[i];
        }, row);
      }

      if (this.body === undefined) {
        this.body = [];
      }

      var counter = 0;
      this.body.forEach(function (row) {
        var idValue = getDataFromDotNotation(_this6.idCol, row);
        if ((idValue === undefined || idValue === null) && (row[_this6.idCol] === undefined || row[_this6.idCol] === null)) {
          row[_this6.idCol] = 'smart_' + counter++;
        }
      });

      var smartBody = [];

      this.body.forEach(function (row) {
        var finalRow = {};
        _this6.tableHeader.forEach(function (col) {
          var realColValue = {};
          if (/\+/.test(col.key)) {
            col.key.split('+').forEach(function (d) {
              realColValue[d] = getDataFromDotNotation(d, row);
            });
          } else {
            realColValue = getDataFromDotNotation(col.key, row);
          }
          finalRow[col.key] = realColValue;
        });
        var idValue = getDataFromDotNotation(_this6.idCol, row);
        var altIdValue = row[_this6.idCol];
        finalRow._id = String(idValue || altIdValue);
        smartBody.push(finalRow);
      });

      this.filteredRowsById = {};
      this.canFilterBy.forEach(function (col) {
        _this6.filterRows(smartBody, _this6.filterAll, col);
      });
      if (this.orderKey !== undefined) {
        smartBody = this.sortRows(smartBody, this.orderKey, this.reverseOrder);
      }

      return smartBody;
    }": TypeError: Cannot read property 'toLowerCase' of undefined (found in component: <smart-table>)warn @ vue.common.js?e881:1019Watcher.get @ vue.common.js?e881:3176Watcher.evaluate @ vue.common.js?e881:3356computedGetter @ vue.common.js?e881:8035smartBody @ SmartTable.vue?4d8e:274Watcher.get @ vue.common.js?e881:3173Watcher.evaluate @ vue.common.js?e881:3356computedGetter @ vue.common.js?e881:8035(anonymous function) @ VM15840:3Watcher.get @ vue.common.js?e881:3173Watcher @ vue.common.js?e881:3158Directive._bind @ vue.common.js?e881:8345linkAndCapture @ vue.common.js?e881:6928compositeLinkFn @ vue.common.js?e881:6897Vue._compile @ vue.common.js?e881:8611Vue.$mount @ vue.common.js?e881:9446Vue._init @ vue.common.js?e881:2468Vue._init @ vue-router.js?e71f:1756VueComponent @ VM15796:2build @ vue.common.js?e881:5837mountComponent @ vue.common.js?e881:5754(anonymous function) @ vue.common.js?e881:5716(anonymous function) @ vue.common.js?e881:5734cb @ vue.common.js?e881:367Vue._resolveComponent @ vue.common.js?e881:8869resolveComponent @ vue.common.js?e881:5736setComponent @ vue.common.js?e881:5715bind @ vue.common.js?e881:5675Directive._bind @ vue.common.js?e881:8325linkAndCapture @ vue.common.js?e881:6928compositeLinkFn @ vue.common.js?e881:6897Vue._compile @ vue.common.js?e881:8611Vue.$mount @ vue.common.js?e881:9446Vue._init @ vue.common.js?e881:2468Vue._init @ vue-router.js?e71f:1756DashboardHomeView @ VM15790:2build @ vue.common.js?e881:5837activate @ vue-router.js?e71f:1176afterData @ vue-router.js?e71f:1232afterActivate @ vue-router.js?e71f:1248activate @ vue-router.js?e71f:1258(anonymous function) @ vue-router.js?e71f:1480step @ vue-router.js?e71f:1503(anonymous function) @ vue-router.js?e71f:1506deactivate @ vue-router.js?e71f:1119step @ vue-router.js?e71f:1505runQueue @ vue-router.js?e71f:1500(anonymous function) @ vue-router.js?e71f:1464step @ vue-router.js?e71f:1503(anonymous function) @ vue-router.js?e71f:1506(anonymous function) @ vue-router.js?e71f:1098(anonymous function) @ vue-router.js?e71f:845Vue._resolveComponent @ vue.common.js?e881:8869resolveAsyncComponent @ vue-router.js?e71f:843canActivate @ vue-router.js?e71f:1090step @ vue-router.js?e71f:1505(anonymous function) @ vue-router.js?e71f:1506(anonymous function) @ vue-router.js?e71f:1098(anonymous function) @ vue-router.js?e71f:845Vue._resolveComponent @ vue.common.js?e881:8869resolveAsyncComponent @ vue-router.js?e71f:843canActivate @ vue-router.js?e71f:1090step @ vue-router.js?e71f:1505runQueue @ vue-router.js?e71f:1500(anonymous function) @ vue-router.js?e71f:1463step @ vue-router.js?e71f:1503(anonymous function) @ vue-router.js?e71f:1506canDeactivate @ vue-router.js?e71f:1073step @ vue-router.js?e71f:1505runQueue @ vue-router.js?e71f:1500start @ vue-router.js?e71f:1462startTransition @ vue-router.js?e71f:2578step @ vue-router.js?e71f:1503(anonymous function) @ vue-router.js?e71f:1506next @ vue-router.js?e71f:1580nextWithBoolean @ vue-router.js?e71f:1591callHook @ vue-router.js?e71f:1632(anonymous function) @ vue-router.js?e71f:2586step @ vue-router.js?e71f:1505runQueue @ vue-router.js?e71f:1500_match @ vue-router.js?e71f:2584onChange @ vue-router.js?e71f:2168listener @ vue-router.js?e71f:979start @ vue-router.js?e71f:982start @ vue-router.js?e71f:2351(anonymous function) @ main.js?3479:196(anonymous function) @ VM15072:310(anonymous function) @ app.js:2425__webpack_require__ @ app.js:556fn @ app.js:87(anonymous function) @ app.js:588__webpack_require__ @ app.js:556(anonymous function) @ app.js:579(anonymous function) @ app.js:582
vue.common.js?e881:1019 [Vue warn]: Error when evaluating expression "function smartBody() {
      var _this5 = this;

      return this.processedSmartBody.map(function (row) {
        var cols = _this5.tableHeader.map(function (col) {
          return col.key;
        });
        cols = cols.filter(function (col) {
          return col !== '_id' || _this5.shouldShowId;
        });
        return { _id: row._id, cols: cols };
      });
    }": TypeError: Cannot read property 'map' of undefined (found in component: <smart-table>)`

**Comment by lavezzi1** [comment]
It crashes when I added the cpu cols in array.

My code:
`    <smart-table
        :body="body"
        :order-by="['img', 'name', 'cpu', 'memory', 'disk', 'region']"
        :can-filter-by="['name.caption', 'memory', 'disk', 'region']"
        >
        <server-img slot="img"></server-img>
        <server-name slot="name"></server-name>
        <server-view-more slot="more"></server-view-more>
    </smart-table>`

**Comment by lavezzi1** [comment]
I also want to know why I get '&#9888' when I add col like this 'created by': 'User'?

**Comment by gurghet** [comment]
As I said, in the /dev branch the filter function is still in development, it doesn't support non-string columns. As a workaround you can write `cpu: '4'` with the 4 around quotes.
`name.caption` is sorting inside a component, for which I'm afraid you will have to wait at least next week, when I plan to finish this feature. But I'll take you use case into consideration.
About columns in the form `'created by': 'User'`. Do you mean in the `:header`, or as part of the body?

**Comment by lavezzi1** [comment]
> About columns in the form 'created by': 'User'. Do you mean in the :header, or as part of the body?

Part of the body, just write this code and get this error:
```
    data: function() {
        return {
            search: '',
            body: [
                {
                    created: '7 days ago',
                    'created by': 'User First'
                }
            ]
        }
    }
```

**Comment by gurghet** [comment]
Yep, you are right. It is a regression from a previous version, I'll look into it!

**Comment by lavezzi1** [comment]
> name.caption is sorting inside a component, for which I'm afraid you will have to wait at least next week, when I plan to finish this feature. But I'll take you use case into consideration.

Cool. It's a little too long time to wait. Can you recommend some table components with key feauture for me – nested component in the columns. I only found your table got this functionality. Thanks.

**Comment by gurghet** [comment]
Actually the reason I created this component is because I din't found any. If you find one please inform me, I'd be interested.
Anyways, turns out spaces are interpreted as separate classes, I'll fix this issue in a next release.

**Comment by lavezzi1** [comment]
@gurghet How search works in the last 2.6.0 beta version?

**Comment by gurghet** [comment]
Have you tried https://github.com/gurghet/vue-smart-table/wiki/Filtering ?

**Comment by lavezzi1** [comment]
Thanks, now it works. My search-input components has a clear button with method:
```
        clearValue() {
            this.value = '';
        },
```
This method removes the value of input but the smart-table doesn't display all data, only what I found. How to reset the search?

**Comment by gurghet** [comment]
Editing inputs programmatically will not trigger related DOM events. Also, you should not rely on long chains of reactivity.
Besides, `value` is the model of the search textbox, you should not rely on it to indirectly send events to the smart-table. Just pretend it doesn't exist.

**Comment by lavezzi1** [comment]
@gurghet So it's not possible to do that action?

**Comment by gurghet** [comment]
I said the exact opposite. You need something like

    clearValue() {
      this.value = '' // <- this won't trigger DOM action which in turn trigger other actions
      this.$broadcast('filter', {filter:'', column:['whatever','columns']})
    },

**Comment by lavezzi1** [comment]
@gurghet Doesn't work...

**Comment by gurghet** [comment]
It's the same event you launch to edit the filter. This is just fixed on the filter value `''`, if it works in the watches it will also work in a method.

**Comment by lavezzi1** [comment]
@gurghet what exactly should I write 
```
    methods: {
        clearValue() {
            // this.value = '';
            this.$broadcast('filter', {filter:'', column:['whatever','columns']})
        },
```

all columns?

**Comment by gurghet** [comment]
Just copy the filter that is working, and paste it with an empty string as a filter.
Also there is no need to comment `this.value = '';`

**Comment by gurghet** [comment]
@lavezzi1 did you have success with the last suggestion?

**Comment by lavezzi1** [comment]
@gurghet no. Can you provide an real example? Say I have two columns: name and number. How this string will looks like?

**Comment by gurghet** [comment]
You would write `this.$broadcast('filter', {filter: '', col:['name', 'number']}`
here's a working jsfiddle: https://jsfiddle.net/gurghet/xpp80yx0/5/

**Comment by lavezzi1** [comment]
Oh, thanks!
